src/ec2/model.ts

```diff
@@ -4,29 +4,102 @@
+        return new Ec2Client(this.regionCode)
+    }
+
+    protected createIamSdkClient(): DefaultIamClient {
+        return new DefaultIamClient(this.regionCode)
+    }
+
+    protected async showConnectError(
+        errorName: Ec2ConnectErrorName,
+        params: Ec2ConnectErrorParameters
+    ): Promise<string> {
+        switch (errorName) {
+            case 'instanceStatus':
+                return (await vscode.window.showErrorMessage(params.message))!
+            case 'permission':
+                return (await showMessageWithUrl(params.message, params.url!, params.urlItem!, 'error'))!
+        }
+    }
+
+    protected async getAttachedPolicies(instanceId: string): Promise<IAM.attachedPoliciesListType> {
+        try {
+            const IamRole = await this.ec2Client.getAttachedIamRole(instanceId)
+            const iamResponse = await this.iamClient.listAttachedRolePolicies(IamRole!.Arn!)
+            return iamResponse.AttachedPolicies!
+        } catch (err) {
+            return []
+        }
+    }
+
+    public async hasProperPolicies(instanceId: string): Promise<boolean> {
+        const attachedPolicies = (await this.getAttachedPolicies(instanceId)).map(policy => policy.PolicyName!)
+        const requiredPolicies = ['AmazonSSMManagedInstanceCore', 'AmazonSSMManagedEC2InstanceDefaultPolicy']
+
+        return requiredPolicies.every(policy => attachedPolicies.includes(policy))
+    }
+
+    public async handleStartSessionError(err: AWSError, selection: Ec2Selection): Promise<string> {
+        const isInstanceRunning = await this.ec2Client.isInstanceRunning(selection.instanceId)
+        const generalErrorMessage = `Unable to connect to target instance ${selection.instanceId} on region ${selection.region}. `
+        const hasProperPolicies = await this.hasProperPolicies(selection.instanceId)
+
+        if (!isInstanceRunning) {
+            const errorParams: Ec2ConnectErrorParameters = {
+                message:
+                    generalErrorMessage +
+                    'Please ensure the target instance is running and not currently starting, stopping, or stopped.',
+            }
+            return await this.showConnectError('instanceStatus', errorParams)
+        }
+
+        if (!hasProperPolicies) {
+            const errorParams: Ec2ConnectErrorParameters = {
+                message:
+                    generalErrorMessage +
+                    'Please ensure the IAM role attached to the instance has the proper policies.',
+                url: 'https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-getting-started-instance-profile.html',
+                urlItem: 'See Policies needed for SSM',
+            }
+            return await this.showConnectError('permission', errorParams)
+        }
 
-        throw new ToolkitError('Start Session Failed. ')
+        throw new ToolkitError('Unknown error unencountered when attempting to start session.', err)
```

Comment:
for this kind of error a `.code` maybe important for telemetry. The choice of `.code` is arbitrary

Reply:
is this where we should mention:

> check that ssm agent is running on the service, and see also the [docs](https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-getting-started.html)”